--- a/src/hotspot/os/posix/os_posix.cpp
+++ b/src/hotspot/os/posix/os_posix.cpp
@@ -74,7 +74,9 @@
 #include <sys/wait.h>
 #include <time.h>
 #include <unistd.h>
+#ifndef __ANDROID__
 #include <utmpx.h>
+#endif

 #ifdef __APPLE__
   #include <crt_externs.h>
@@ -453,6 +455,7 @@ void os::Posix::print_load_average(outputStream* st) {
 // unfortunately it does not work on macOS and Linux because the utx chain has no entry
 // for reboot at least on my test machines
 void os::Posix::print_uptime_info(outputStream* st) {
+#ifndef __ANDROID__
   int bootsec = -1;
   int currsec = time(nullptr);
   struct utmpx* ent;
@@ -467,6 +470,7 @@ void os::Posix::print_uptime_info(outputStream* st) {
   if (bootsec != -1) {
     os::print_dhm(st, "OS uptime:", (long) (currsec-bootsec));
   }
+#endif
 }

@@ -2030,6 +2034,16 @@
   // Run the command with /bin/sh to parse the command string.
   char* const argv[] = { (char*) "sh", (char*) "-c", (char*) cmd, nullptr };
   char* const env[] = { nullptr };
+#if defined(__ANDROID__) && __ANDROID_API__ < 28
+  pid_t pid = fork();
+  if (pid == 0) {
+    execve("/bin/sh", argv, env);
+    _exit(127);
+  }
+  if (pid > 0) {
+    int status;
+    waitpid(pid, &status, 0);
+  }
+#else
   int rc = ::posix_spawn(&pid, "/bin/sh", nullptr, nullptr, (char**) argv, env);
+#endif
