diff --git a/src/hotspot/os/linux/os_perf_linux.cpp b/src/hotspot/os/linux/os_perf_linux.cpp
index 0000000..1111111 100644
--- a/src/hotspot/os/linux/os_perf_linux.cpp
+++ b/src/hotspot/os/linux/os_perf_linux.cpp
@@ -52,6 +52,10 @@
 #include <sys/sysinfo.h>
 #include <sys/times.h>

+#if defined(__ANDROID__)
+#define gethostbyname_r(a,b,c,d,e,f) (-1)
+#endif
+
 static const char MNT_DIR[] = "/proc/mounts";
 static const char SYSTEM_BUS_SYS_DEVICES_SYSTEM_CPU[] = "/sys/devices/system/cpu";
